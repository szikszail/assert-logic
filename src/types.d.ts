
export type AnyValue = void | boolean | string | object | number | null;
export type AsyncAnyValue = Promise<AnyValue>;
export type AssertionFunction = () => AnyValue | AsyncAnyValue;
export type AssertionValue = AssertionFunction | AnyValue | AsyncAnyValue;
