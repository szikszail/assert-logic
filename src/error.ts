import {Assertion} from "./assertion/assertion";
import {stringify} from "yaml";
import {AnyValue, EvaluationResult} from "./types";
import {valueToString} from "./assertion/pass";

type ErrorRepr = Record<string, string | (ErrorRepr | string)[]>;

function getErrorRepresentation(operator: string, expectation: string, assertion: Assertion, results?: EvaluationResult[]): ErrorRepr {
  const repr: ErrorRepr = {
    [`AssertionError (${operator})`]: expectation,
  };
  if (results?.length > 0) {
    repr.Results = results.map((result) => {
      if (result instanceof AssertionError) {
        return getErrorRepresentation(result.operator, result.expectation, result.assertion, result.results);
      }
      if (result instanceof Error) {
        return {
          // eslint-disable-next-line no-control-regex
          Error: result.message.replace(/\u001b\[\d+m/g, ''),
        };
      }
      return "Pass";
    });
  }
  repr.Expression = assertion.toString();
  return repr;
}

export class AssertionError extends Error {
  constructor(
    public readonly operator: string,
    public readonly expectation: string,
    public readonly assertion: Assertion,
    public readonly results?: EvaluationResult[],
  ) {
    super(stringify(
      getErrorRepresentation(operator, expectation, assertion, results),
      {
        indent: 2,
        // defaultStringType: "PLAIN",
        // defaultKeyType: "PLAIN",
        collectionStyle: 'block',
        // lineWidth: 0,
        // blockQuote: true,
        doubleQuotedMinMultiLineLength: Infinity,
      }
    ));
    this.name = "AssertionError";
  }
}

export class FailedAssertionError extends Error {
  constructor(value: AnyValue) {
    super(`Failed expression: ${valueToString(value)}`);
    this.name = "FailedAssertionError";
  }
}