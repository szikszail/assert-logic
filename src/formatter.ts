import type {AssertionValue} from "./types";

export function toString(value: AssertionValue): string {
  if (value instanceof Function) {
    let name = value?.name;
    if (!name) {
      name = value.toString();
    }
    if (name.length > 20) {
      name = name.slice(0, 17) + "...";
    }
    return `function(${name})`;
  }
  if (value instanceof Promise) {
    return value.toString();
  }
  return `value(${value})`;
}