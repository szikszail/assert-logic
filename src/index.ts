// import debug = require("debug");

// const log = debug("assert-logic");

import type {AssertionValue, AssertionGeneratorFunction} from "./types";
import {PASSAssertion} from "./assertion/pass";
import {NOTAssertion} from "./assertion/not";
import {ORAssertion} from "./assertion/or";
import {ANDAssertion} from "./assertion/and";
import {XORAssertion} from "./assertion/xor";
import {NORAssertion} from "./assertion/nor";
import {NANDAssertion} from "./assertion/nand";
import {XNORAssertion} from "./assertion/xnor";
import {EVERYAssertion} from "./assertion/every";
import {SOMEAssertion} from "./assertion/some";

import {setDefaultOptions} from "lines-builder";

setDefaultOptions({
  // Need to force EOL to be Unix style, because Windows style is not supported by YAML
  eol: "\n",
  trimLeft: false,
  skipFirstLevelIndent: true,
  indent: 2
});

export const pass = (value: AssertionValue) => new PASSAssertion(value);
export const not = (value: AssertionValue) => new NOTAssertion(value);
export const or = (...values: AssertionValue[]) => new ORAssertion(...values);
export const and = (...values: AssertionValue[]) => new ANDAssertion(...values);
export const nor = (...values: AssertionValue[]) => new NORAssertion(...values);
export const nand = (...values: AssertionValue[]) => new NANDAssertion(...values);
export const xor = (...values: AssertionValue[]) => new XORAssertion(...values);
export const xnor = (...values: AssertionValue[]) => new XNORAssertion(...values);
export const every = <T>(items: T[], fn?: AssertionGeneratorFunction<T>) => new EVERYAssertion(items, fn);
export const some = <T>(items: T[], fn?: AssertionGeneratorFunction<T>) => new SOMEAssertion(items, fn);
