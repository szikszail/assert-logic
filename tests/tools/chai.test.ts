import {or, and} from '../../src';
import {expect as chaiExpect, assert as chaiAssert, should } from 'chai';

describe('chai', () => {
  describe('expect', () => {
    test("should work with Chai/expect assertions (positive)", () => {
      const number = 12;
      expect(() => and(
        () => chaiExpect(number).to.be.a('number'),
        () => chaiExpect(number % 2).to.equal(0),
        or(
          () => chaiExpect(number).to.be.lessThan(15),
          () => chaiExpect(number).to.be.greaterThan(10),
        ),
      ).evaluate()).not.toThrow();
    });
    test("should work with Chai/expect assertions (negative)", () => {
      const number = 12;
      expect(() => and(
        () => chaiExpect(number).to.be.a('number'),
        () => chaiExpect(number % 2).to.equal(1),
        or(
          () => chaiExpect(number).to.be.lessThan(7),
          () => chaiExpect(number).to.be.greaterThan(10),
        ),
      ).evaluate()).toThrowErrorMatchingSnapshot()
    });
  });

  describe('assert', () => {
    test("should work with Chai/assert assertions (positive)", () => {
      const number = 12;
      expect(() => and(
        () => chaiAssert.isNumber(number),
        () => chaiAssert.equal(number % 2, 0),
        or(
          () => chaiAssert.isBelow(number, 15),
          () => chaiAssert.isAbove(number, 10),
        ),
      ).evaluate()).not.toThrow();
    });
    test("should work with Chai/assert assertions (negative)", () => {
      const number = 12;
      expect(() => and(
        () => chaiAssert.isNumber(number),
        () => chaiAssert.equal(number % 2, 1),
        or(
          () => chaiAssert.isBelow(number, 7),
          () => chaiAssert.isAbove(number, 10),
        ),
      ).evaluate()).toThrowErrorMatchingSnapshot()
    });
  });

  describe('should', () => {
    beforeAll(() => {
      should();
    });

    test("should work with Chai/should assertions (positive)", () => {
      const number = 12;
      expect(() => and(
        () => number.should.be.a('number'),
        () => (number % 2).should.equal(0),
        or(
          () => number.should.be.lessThan(15),
          () => number.should.be.greaterThan(10),
        ),
      ).evaluate()).not.toThrow();
    });

    test("should work with Chai/should assertions (negative)", () => {
      const number = 12;
      expect(() => and(
        () => number.should.be.a('number'),
        () => (number % 2).should.equal(1),
        or(
          () => number.should.be.lessThan(7),
          () => number.should.be.greaterThan(10),
        ),
      ).evaluate()).toThrowErrorMatchingSnapshot()
    });
  });

});