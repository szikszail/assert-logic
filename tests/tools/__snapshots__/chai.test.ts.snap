// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chai assert should work with Chai/assert assertions (negative) 1`] = `
"AssertionError (AND): Expected all expression to pass, but not all did.
Results:
  - Pass
  - AssertionError (PASS): Expected expression to pass.
    Results:
      - Error: expected +0 to equal 1
    Expression: (function)
  - Pass
Expression: |-
  AND(
    (function)
    (function)
    OR(
      (function)
      (function)
    )
  )
"
`;

exports[`chai expect should work with Chai/expect assertions (negative) 1`] = `
"AssertionError (AND): Expected all expression to pass, but not all did.
Results:
  - Pass
  - AssertionError (PASS): Expected expression to pass.
    Results:
      - Error: expected +0 to equal 1
    Expression: (function)
  - Pass
Expression: |-
  AND(
    (function)
    (function)
    OR(
      (function)
      (function)
    )
  )
"
`;

exports[`chai should should work with Chai/should assertions (negative) 1`] = `
"AssertionError (AND): Expected all expression to pass, but not all did.
Results:
  - Pass
  - AssertionError (PASS): Expected expression to pass.
    Results:
      - Error: expected +0 to equal 1
    Expression: (function)
  - Pass
Expression: |-
  AND(
    (function)
    (function)
    OR(
      (function)
      (function)
    )
  )
"
`;
